test/unit/app/utils/ussd/kenya/test_pin_state.py

```diff
@@ -110,15 +110,16 @@ def test_kenya_state_machine(test_client, init_database, user_factory, session_f
  ])
 def test_authorize_pin(test_client, init_database, session_factory, user_factory, user_input, expected,
                        before_failed_pin_attempts, after_failed_pin_attempts):
-    session = session_factory()
-    user = user_factory(phone='+6140000000')
-    user.failed_pin_attempts = before_failed_pin_attempts
+    with db.session.no_autoflush:
```

Comment:
is there autoflush here?

Reply:
Yes there's a query invoked autoflush there

Reply:
```
$ grep autoflush test -r
test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py:    with db.session.no_autoflush:
Binary file test/unit/app/utils/ussd/kenya/__pycache__/test_kenya_ussd_processor.cpython-36-PYTEST.pyc matches
```

I'm a bit sceptical here. _no other_ test has explicit autoflush check. Why is it needed in _this particular case please.

Reply:
So when I run the test without the autoflush I run into an error along the lines of there being a user existing who bore the same number as the one that piece of code attempts to create.
The error then suggested that the cause of the error was a premature autoflush that was being invoked by some query within the flow of that piece of code. It the suggested that I run it without the autoflush to avoid the error and for sure it worked.

Reply:
the way the other tests work afaiu is that then you make a fixture for a different user.

I do not think we should introduce session-control methods that aren't introduced before. That can make it harder to clean up once we make the pass to harden, sanitize and standardize the tests and fixtures.

Reply:
Makes sense.
I am gonna look at this other or another approach

Reply:
> . I found one in:
test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py in the test_custom_display_text 

which commit? I could not find that in your revision.

Reply:
> > . I found one in:
> > test/unit/app/utils/ussd/kenya/test_kenya_ussd_processor.py in the test_custom_display_text
> 
> which commit? I could not find that in your revision.

https://github.com/GrassrootsEconomics/SempoBlockchain/commit/8517db39912709f00acf2b5d9d1a9741355c8a00

Reply:
```
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 39) def test_custom_display_text(test_client, init_database, menu_name, language, expecteds,
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 40)                              real_at_idx, length, unexpected, menu_nr):
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 41)     with db.session.no_autoflush:
8517db399 (RubenGeo 2019-12-10 08:06:37 +0100 42)         start_state = UssdSessionFactory(state="menu_name")
```

Ok, thanks, I yield.

It seems like a hack, though, so we should make a point of weeding this out. I guess we are adding that to technical debt, unless you already found a way to do this cleanly and painlessly? We really need to get this functionality out.